### ISSUE TITLE
[123] mode does not produce input on any input fields on KYF33

### ISSUE THREAD

**Issue by mcfrei** [sspanak/tt9#241]
Thank you so much for a wonderful project! Could you help me diagnose this issue better?

Short description: entering digits in phone number fields only does not work, digits do not appear.

Longer description: On fields for phone numbers, this keyboard does nothing (does not input a number) when pressing number keys, if field is open. I can enter exactly one digit by moving from field with cursor keys, returning to it and pressing the digit while keyboard is not open.

Device: Kyocera Torque X01 KYF33
Android version: 5.1.1
No touchscreen.

Comments: Some of the fields are [123] only on this device, for example phone numbers (it means that "change Input Mode" key does not do anything in them) (probably InputType.TYPE_CLASS_PHONE or something like this).

Existing IME input does work on this phone, unfortunately does not support any languages except Japanese and English. I can take a video of trying to input digits, or extract the PhoneBook.apk, if that helps.

**Comment by sspanak** [comment]
This may be the same as #225. Please, check out the discussion there.

**Comment by mcfrei** [comment]
I plan to set up Android Studio on weekend to try and help fixing it.
Thank you for a quick answer and a latest beta version!
However, the problem of not entering the numbers in the restricted InputType fields is still present and is still the same in the latest beta version 20.1 (a0b9e09)

**Comment by mcfrei** [comment]
Gladly. However, I added quite a lot of names to my personal dictionary, and I am using the debug build I'm building myself, so I installed a version I built from commit b3fde40aa. I have verified that install was successful by checking the info in the About section of Settings.

Unfortunately, I am still only able to input one digit at a time, by selecting and deselecting the input field.

I am also still receiving a lot of following errors in logcat"

`SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length` 

I now have more experience using this Kyocera phone and what I have noticed is that when I use the preinstalled keyboard to enter numbers in the text fields, another text field is opened over the app window, that takes whole screen. And tt9 does not do that. I wonder is that a standard behavior?

What kind of debug info could be helpful here? Logcat? video of the work, maybe? I would be really grateful for a hint.

**Comment by mcfrei** [comment]
What I have found is that every time I press or release a number button in the calling interface getInstance is called and a new ModeDialer is created, and when the number key is doubled in the calling interface, there are two ModeDialers created at the exact same time.

I have found it by adding debug log in getInstance and various other functions, I can push them if you are interested. 

Regarding initTyping in TraditionalT9 - there is an init in determineAllowedInputModes and reinitlater when you do the getInstance, so I assume that in Kyocera calling interface input is reinit with every key press and release.

**Comment by mcfrei** [comment]
Here are logcat logs from the first button press, when the digit [4] IS inserted in the "phone number field":

```
2023-05-20 20:43:28.662 11367-11367 tt9/onStartInputView    io.github.sspanak.tt9                D  inputType: 3 fieldId: 19 fieldName: null packageName: com.android.contacts restarting: false
2023-05-20 20:43:28.702 11367-11367 onKeyDown               io.github.sspanak.tt9                D  Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_4, scanCode=5, metaState=0, flags=0x8, repeatCount=0, eventTime=7012440, downTime=7012440, deviceId=23, source=0x301 } repeat?: 0 long-time: false
2023-05-20 20:43:28.952 11367-11367 onKeyUp                 io.github.sspanak.tt9                D  Key: 11 repeat?: 0
```

And here are logcat logs from the second button press, when the digit [5] is NOT inserted in the "phone number field":

```
2023-05-20 20:46:28.142 11367-11367 onKeyDown               io.github.sspanak.tt9                D  Key: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_5, scanCode=6, metaState=0, flags=0x8, repeatCount=0, eventTime=7191920, downTime=7191920, deviceId=23, source=0x301 } repeat?: 0 long-time: false
2023-05-20 20:46:28.262 11367-11367 onKeyUp                 io.github.sspanak.tt9                D  Key: 12 repeat?: 0
```

They seem the same so I will continue debugging.

**Comment by mcfrei** [comment]
Regarding k9t9 - I have installed the latest release from github:
https://github.com/danielshorten/k9t9/releases/tag/0.12.0
and I can confirm that it does allow input correctly in "phone number fields" in com.android.contacts, not ignoring key presses when field is "in focus".

**Comment by mcfrei** [comment]
I've narrowed it down to `sendDownUpKeyEvents(mInputMode.getKeyCode());` not producing a number in "phone number fields" in `TraditionalT9.handleSuggestions()`

**Comment by mcfrei** [comment]
I also verified that `sendDownUpKeyEvents(mInputMode.getKeyCode());` is the exact line that produces first number when field was not focused and keyboard was not open, by commenting that line and checking that I no longer can enter one digit by leaving field and returning to it and by replacing `mInputMode.getKeyCode()` in it with `KeyEvent.KEYCODE_8` and verifying that the digit is always 8 even if I press other key. I am still puzzled why `sendDownUpKeyEvents` stops working after the keyboard is open in `[123]` mode though. The search continues.

**Comment by sspanak** [comment]
>I am still puzzled why sendDownUpKeyEvents stops working after the keyboard is open in [123] mode though. The search continues.

It may have to do with the restarts you are experiencing. `currentInputConnection` probably becomes invalid, so no key events are making through. This is just a theory, though.

**Comment by mcfrei** [comment]
I have noticed that [123] mode actually does not work in any input fields, not only in numeric, both in in b3fde40 and in 4405c0e.
But there is one curious exception: long-pressing the `0` button does produce the `+` symbol.

**Comment by mcfrei** [comment]
I managed to fix the issue finally by using suggestions instead of keyCode in Mode123/ModeDialer, I'll get a pull request ready for you to check out and test on your device asap.

**Comment by mcfrei** [comment]
Something is wrong with long press, i need more time. :(
Specifically, with my fixes, after entering dialer interface with long pressing the `*` button, no keyboard input passes through, thus invalidating fix for #261

**Comment by mcfrei** [comment]
Issue is still present in  09e5e1b (v 22.0), but the cause for it is that 
`sendDownUpKeyEvents(mInputMode.getKeyCode());` in `handleSuggestions` does not seem to work at all on KYF33, so I was thinking that I could make a Mode123Alt class that uses `suggestions` instead of the `keyEvent` but I am not sure how to set it up so that it only is being used on KYF33, probably like that in `InputMode.getInstance` ?
```
case MODE_123:
   if(settings.is123AltMode())
      return new Mode123();
   else
     return new Mode123Alt();
```

**Comment by mcfrei** [comment]
See also my take on solving this idea in https://github.com/sspanak/tt9/compare/master...mcfrei:tt9:master

**Comment by sspanak** [comment]
There may be a simpler way of doing what you are trying to do.

It seems that Mode123 is working fine on your phone, if it produces text instead of key codes. But this is weird, because other modes also produce key codes sometimes. So I need to know the following:

1. Make sure to "#" or "*" are not hotkeys, so you can use them for typing. Do they work in ABC or Predictive modes? If you go back to 123 from another mode, do they still work? Are backspace and OK working. They also send actions using `sendDownUpKeyEvents()`, so if there is something wrong with that function, they must be affected too.
2. If the number keys are not working, but Backspace and OK are fine, maybe the key codes are wrong. In `TraditionalT9.handleSuggestions()`, what does `mInputMode.getKeyCode()` return? Are the key codes equal to `KeyEvent.KEYCODE_0` through `KeyEvent.KEYCODE_9`, `KeyEvent.KEYCODE_POUND` and whatnot?
3. If the key codes are fine, you could change the way they are sent to the system:
  - In `Mode123.onOtherKey()`, remove the conditions, just do:
```
	public boolean onOtherKey(int key) {
		reset();
		keyCode = key;
		return true;
	}
```
  - Find `TraditionalT9.handleSuggestions()`
  - Change `sendDownUpKeyEvents(mInputMode.getKeyCode())` to this:
```
	currentInputConnection.sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, mInputMode.getKeyCode()));
    	currentInputConnection.sendKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, mInputMode.getKeyCode()));
```
4. If nothing else works, you could make `Mode123` return strings:
 - Change `onNumber()` and `reset()` like this:
```
	@Override
	public boolean onNumber(int number, boolean hold, int repeat) {
		reset();
		suggestions.add((number == 0 && hold) ? "+" : String.valueOf(number));
		return true;
	}

	@Override
	public void reset() {
		super.reset();
		autoAcceptTimeout = 0;
	}
```
In that case `onOtherKey()` will be a bit more complex, but we'll get to it if needed.

Please, try all the scenarios above and let me know of the results. It is important not to do big changes to the Numeric mode, not to break it on other phones.

**Comment by mcfrei** [comment]
Thank you so much for the review and the ideas on how to do it better! Here are the results of the tests.

All tests were conducted on last master cf76633 (verified by checking the About info in settings, very handy!).

1. I removed `#` and `*` from hotkeys, and they can't type the corresponding symbol in any of the modes, namely `[ abc / en ]`, `[ english ]`, or `[ 123 ]`. Backspace deletes the last symbol both when it's bind on `Back` key, and if it's bind on `F3` key. Ok also works both by exiting the "choose word" mode and after that by switching to the next input field.

2. I have used the Logger.i to output the code that is being returned. The key codes are indeed equal to their constants in `KeyEvent' class, with `0` equal to 7, `*` to 17 and `#` to 18.

3. I have tested this approach (replacing `sendUpDownKeyEvents` with two `sendKeyEvent` earlier and it did not work, but I did test it again exactly as you suggested and they definitely do not produce the corresponding symbol in the text field. I did not find any other ways to send key event, so I tried using suggestions.

4. Exactly that was what I made originally, and after the code became too complicated, I separated the strings functionality to Mode123, aiming to choose between `new Mode123` and `new Mode123Alt` in `InputMode.getInstance` either by determining device model somehow, or by changing the preferred mode in settings. The full functionality also required adding `boolean hold` to `onOtherKey` to support entering alternative `*` and `#` characters on hold, correspondingly `+` and `;`.

**Comment by sspanak** [comment]
So, generating numeric strings is the way to go. I reviewed k9t9 and it looks like it is doing the same.

> Exactly that was what I made originally, and after the code became too complicated, I separated the strings functionality...

As I said, there is a simpler way of doing what you want to do. Let's try this solution:

```
package io.github.sspanak.tt9.ime.modes;

import android.view.KeyEvent;

import androidx.annotation.NonNull;

public class Mode123 extends ModePassthrough {
	@Override public int getId() { return MODE_123; }
	@Override @NonNull public String toString() { return "123"; }

	@Override public final boolean is123() { return true; }
	@Override public boolean isPassthrough() { return false; }

	@Override
	public boolean onNumber(int number, boolean hold, int repeat) {
		reset();
		suggestions.add((number == 0 && hold) ? "+" : String.valueOf(number));
		return true;
	}

	@Override
	public boolean onOtherKey(int key) {
		reset();

		int keyChar = new KeyEvent(KeyEvent.ACTION_DOWN, key).getUnicodeChar();
		if ((keyChar > 31 && key < 65) || (keyChar > 90 && keyChar < 97) || (keyChar > 122 && keyChar < 127)) {
			suggestions.add(String.valueOf((char) keyChar));
			return true;
		}

		return false;
	}

	@Override
	public void reset() {
		super.reset();
		autoAcceptTimeout = 0;
	}
}

```

That's the entire `Mode123.java`. It is working fine on both my phones and I hope it would work on yours too. If it does, I guess it would be safe to assume it would work on any phone. :smile:

**Comment by owen-young** [comment]
Thank you for pointing out this issue to me. I should have searched through existing issues before making my own, I apologize. That code you posted did work on my phone though! 123 mode is working well. Thanks a bunch!

**Comment by sspanak** [comment]
That's great, thanks for confirming. It's really hard to debug these when I don't have a Kyocera phone. I suppose I can add this in the next version or the one after.

**Comment by owen-young** [comment]
Thanks! One quirk I have noticed is when in 123 mode, the # key does a new line while also changing the mode. Not sure if this is intended or not, I can look into it more when I get home from work.

-------- Original Message --------

**Comment by sspanak** [comment]
Definitely, it's not intended. Probably, the default action on your phone is new line and it is getting through instead of being suppressed. Have in mind, I wrote and tested the code above in 10 minutes or so. It is not fully tested, so I may have missed something. I'll double check when I get back to this issue. Thanks for reporting!

**Comment by mcfrei** [comment]
I'll get a pull request ready with that functionality.  I assumed that sendDownUpKeyEvents could be crucial for some functionality on other phones, and I put more effort in long pressing `*` and `#` keys, I'll test your solution to see if it produces the same result.
What is the result of long key press of `#` and `*` keys on dialer on your devices, @sspanak ?

**Comment by mcfrei** [comment]
Thanks so much for providing your test results promptly, I am very grateful for them!

Apparently, `,` is a "pause slightly" functionality, and `;` is a "wait for response and continue" functionality in phone numbers, although I don't know how popular it is outside Japan.

There is a [E.123](https://en.wikipedia.org/wiki/E.123) standard article, where the most interesting info parts are that 
- in E.123 A tilde (~) indicates an additional [dial tone](https://en.wikipedia.org/wiki/Dial_tone) that the user should wait for.
- in Microsoft Number Format the calling sequence can contain dial-able numbers such as digits `0–9` and DTMF tones `ABCD*#`, formatting characters `␣ . -`, and control characters `! P T , W @ $ ? ;`, which correspond to the Dial command of the Hayes [AT command set](https://en.wikipedia.org/wiki/AT_command_set).

There is also another [link](https://www.macsparky.com/blog/2015/04/the-iphone-extension-trick/) on the topic of `;` usage in iPhones.

Considering all this, I would still prefer implementing long press on symbol characters `#` and `*` (as `;` and `+` correspondingly) in both Dialer and 123  mode. 

Here are my test results:
On KYF33, using default preinstalled input `iWnn IME` long press is as follows: 
- in dialer, long press `#` is `;', long press `*` is `+` (in any place of the number, not only the first one), and long press `0` produces repeating zeroes like other number keys.
- not in dialer number inputs (in com.android.contacts phone number fields) the long press functionality is the same except that long pressing number keys does not produce repeating digits.
- Remark: default preinstalled input opens a full screen input when editing starts, and that is convenient because the phone screen is relatively small, and because entering Japanese characters usually involves heavy character replacement and suggestions (see also: hiragana, katakana and kanji).

Considering all that plus your information, I would suggest implementing `+` as a character for long pressing both `0` and `*` keys, to provide maximum compatibility for you, me and other users. Alternatively I could add options for "long press" in the "Keypad" area of the settings. consisting of "long press 0", "long press *" and "long press #" entries. 

Considering the #225 bug, that is the reason why I implemented Mode123Alt in my commit as a separate mode, intending having an option for users to choose between it and original Mode123. Your wonderful idea about checking the original k9keboard could help to debug the issue for @lgexalter too, if they are still interested in it.

Regarding k9 keyboard, it seems that it uses inputConnection composing text feature extensively by calling setComposingText and finishComposing even in Dialer input.

**Comment by mcfrei** [comment]
I was able to check last version 4edda7c on KYF33 and here are the test results:
- dialer app works perfectly, no character doubling and long presses work;
- other apps - behaviour is the same, no input in numeric fields in [123] mode.

I worked on a pull request but encountered unexpected behaviour - on KYF33 when a `*` or `#` key is long pressed, there is no `onKeyLongPress` event in `KeyPadHandler`, and instead an `onKeyDown` event with `event.isLongPress() == true` is produced.

I was able to fix that hastily, but I don't like the way I did it in [KYF33-2](https://github.com/mcfrei/tt9/tree/KYF33-2) branch.

Is that an expected behavior? How can one deal with it, considering differences in processing logic in `onKeyDown` and `onKeyLongPress`? Have you seen stuff like this, @sspanak ?

**Comment by sspanak** [comment]
> Apparently, `,` is a "pause slightly" functionality, and `;` is a "wait for response and continue" functionality in phone numbers, although I don't know how popular it is outside Japan.

I've never had these on any phone, even in the old Nokia days. Neither have I ever seen them on a touchscreen dialer. I have had only Android phones though, not an iPhone. Indeed, carriers may not support such control codes in my part of the world.

> Regarding k9 keyboard, it seems that it uses inputConnection composing text feature extensively by calling setComposingText and finishComposing even in Dialer input.

But this is causing double numbers on my phone every now and then. It seems that the system and the IME are in a constant race and sometimes they both win and type a number.

> I worked on a pull request but encountered unexpected behaviour - on KYF33 when a * or # key is long pressed, there is no onKeyLongPress event in KeyPadHandler, and instead an onKeyDown event with event.isLongPress() == true is produced.

Ok, now I realized what the problem is. No matter what, `*` and `#` are always handled in `KeyPadHandler.onKeyDown` (we are returning `true` when they are pressed). But this would prevent the system `onKeyLongPress` from working, which is a problem on your phone, because there are actions assigned to long pressing these keys.

The above also means `ModePassthrough` is not passing through these keys, while its purpose is to provide maximum compatibility on all phones, so their "special" functions would work as expected. And this is where I would like to go. I'd rather allow the phone process key presses whenever possible, instead of writing code that does the same. Not only that, but have per-device implementations. This is what I said I don't like in my previous comment.

I admit I overlooked the case with `*` and `#` in `ModePassthrough` and it needs to be fixed. Since I am currently busy with adding new languages and you are on the topic, could you try one simple thing?

1. Use the latest `master`
2. Perhaps, use the new `Mode123` that I suggested above.
3. In `KeyPadHandler.onKeyDown()` find the `if` marked with this comment `// start tracking key hold`. Add `Key.isPoundOrStar(keyCode)` to the condition and see if this fixes the problem for you?

I see you did something very similar in your branch, but I think all else is not really necessary.

> see also: hiragana, katakana and kanji

日本語を少し読めます :slightly_smiling_face:

**Comment by sspanak** [comment]
As for the AT and Microsoft command set. Well... let's not go there. We are not implementing a software modem here, right? :slightly_smiling_face:

**Comment by sspanak** [comment]
@mcfrei, @owen-young, the latest [version 23.7](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-23.7.apk) fixes a lot of small bugs. Could you please try if it fixes the problem for you?

If you prefer to build from source, please use the `mode-refactoring-again` branch. And if the branch is gone by the time you read this, then it has been merged to `master`, so use `master` instead.

**Comment by felipecrs** [comment]
Unfortunately the exact same issue remains with 23.7. :(

**Comment by sspanak** [comment]
#345 is somewhat related to this. In the end, I have added proper support for all possible characters in 123 mode, depending on the input field (phone number or plain numeric). 

I believe all outstanding problems discussed here will be fixed in the upcoming version 26.0, but I'll leave the issue open, for anyone affected to confirm, because I can't do it on my phone.

**Comment by sspanak** [comment]
If anyone is still experiencing this, please try out [v27.0](https://github.com/sspanak/tt9/releases/tag/v27.0).

If no one comments after this, I'll consider the problem to be resolved.
